File: Cpp/sources/heap.cpp
#include <iostream>
#include <ctime>
#include "../headers/swap.h"
#include "../headers/value.h"

using namespace std;

void max_heapify(int* a, int s, int i)
{
    int l = i;
    int ln = 2 * i + 1;
    int rn = 2 * i + 2;

    if (ln < s && a[ln] > a[l])
        l = ln;
    if (rn < s && a[rn] > a[l])
        l = rn;

    if (l != i)
    {
        swap(a, l, i);
        max_heapify(a, s, l);
    }
}

void heap_sort(int* a, int s)
{
    for (int i=s/2-1;i>-1;--i)
        max_heapify(a, s, i);
    for (int i=s-1;i>0;--i)
    {
        swap(a, i, 0);
        max_heapify(a, i, 0);
    }
}

int main()
{
    int s = 1000000; // Changeable size
    int *a = rev_p(s);

    clock_t start = clock();

    heap_sort(a, s);

    clock_t ended = clock();
    double total = (double)(ended - start) / CLOCKS_PER_SEC;

    for (int i=0;i<s;++i)
        cout << a[i] << " " << endl;
    cout << total << " seconds";

    free(a);
    return 0;
}

File: Python/sources/heap.py
import time
from Python.Python.others.swap import *
from Python.Python.others.value import *


def max_heapify(a, s, i):
    l = i
    ln = 2 * i + 1
    rn = 2 * i + 2

    if ln < s and a[ln] > a[l]:
        l = ln
    if rn < s and a[rn] > a[l]:
        l = rn
    if l != i:
        swap(a, l, i)
        max_heapify(a, s, l)

def heap_sort(a, s):
    for i in range(s//2-1, -1, -1):
        max_heapify(a, s, i)
    for i in range(s-1, 0, -1):
        swap(a, i, 0)
        max_heapify(a, i, 0)


# Changeable size
s = 1000000
a = rev_p(s)

start = time.time()

heap_sort(a, len(a))

end = time.time() - start

print(a)
print(end, "seconds")
